File: test/opdecode.cc
#include <assert.h>
#include <err.h>
#include <stdio.h>

#include "spice_load.h"

enum alu_ops_t {
    op_add = 0,
    op_sub = 4,
    op_or  = 8,
    op_and = 12,
    op_adc = 16,
    op_sbc = 20,
    op_xor = 24,
    op_dnc = 28,
    op_alu_mask = 28
};

enum regreg_t {
    op_inc  = 192 + 0,
    op_dec  = 192 + 4,
    op_move = 192 + 8,
    op_load = 192 + 12,
    regreg_mask = 0xcc,
};

static const char * const alu_ops[] = {
    "ADD", "SUB", "OR", "AND", "ADC", "SBC", "XOR", "DNC" };

int main(int argc, char * argv[])
{
    spice_load S(argc, argv, 5e-6);

    const auto I2 = S.extract_signal("i2");
    const auto I3 = S.extract_signal("i3");
    const auto I4 = S.extract_signal("i4");
    const auto I5 = S.extract_signal("i5");
    const auto I6 = S.extract_signal("i6");
    const auto I7 = S.extract_signal("i7");
    const auto CO = S.extract_signal("co");

    const auto CR = S.extract_signal("cr");
    const auto CSi = S.extract_signal("cs#");
    const auto CoEi = S.extract_signal("coe#");
    const auto ARi = S.extract_signal("ar#");
    const auto AS = S.extract_signal("as");
    const auto AND = S.extract_signal("and");
    const auto OR = S.extract_signal("or");
    const auto Ni = S.extract_signal("n#");

    const auto INi = S.extract_signal("in#");
    const auto OUT = S.extract_signal("out");
    const auto MW = S.extract_signal("mw");
    const auto MRi = S.extract_signal("mr#");

    const auto CWi = S.extract_signal("cw#");

    for (int i = 0; i != S.num_samples; ++i) {
        int opcode = I7[i] * 128 + I6[i] * 64 + I5[i] * 32
            +        I4[i] * 16  + I3[i] * 8 + I2[i] * 4;
        bool co = CO[i];

        bool cr = CR[i];
        bool cs = !CSi[i];
        bool coe = !CoEi[i];

        bool ar = !ARi[i];
        bool as = AS[i];

        bool And = AND[i];
        bool Or = OR[i];
        bool n = !Ni[i];

        bool in = !INi[i];
        bool out = OUT[i];
        bool mw = MW[i];
        bool mr = !MRi[i];
        bool cw = !CWi[i];

        const char * tag = "";
        bool ex_coe = co;
        bool ex_cr = false;
        bool ex_cs = false;
        bool ex_ar = false;
        bool ex_as = false;
        bool ex_and = false;
        bool ex_or = false;
        bool ex_n = false;
        bool ex_in = false;
        bool ex_out = false;
        bool ex_mr = false;
        bool ex_mw = false;
        bool ex_cw = false;
        bool check_alu = false;

        // Instructions that use ALU output value; ALU ops + INC/DEC/MV
        if (opcode >= 0x80 && (opcode & 0xcc) != 0xcc)
            check_alu = true;

        // Instructions that write the C flag: ALU ops plus CMP/TST.
        if ((opcode >= 0x80 && opcode <= 0x9f)
            || (opcode & 0xf4) == 0x64) {
            check_alu = true;
            ex_cw = true;
        }

        if (opcode < 0xc0) {
            switch (opcode & op_alu_mask) {
            case op_add:
                ex_coe = false;
                break;
            case op_sub:
                ex_coe = true;
                ex_n = true;
                break;
            case op_adc:
                break;
            case op_sbc:
                ex_n = true;
                break;
            case op_or:
                ex_coe = false;
                ex_cr = true;
                ex_or = true;
                break;
            case op_xor:
                ex_coe = false;
                ex_cr = true;
                break;
            case op_and:
                ex_coe = true;
                ex_cs = true;
                ex_and = true;
                break;
            case op_dnc:
                ex_coe = coe;
                ex_cs = cs;
                ex_and = And;
                break;
            }
        }
        switch (opcode & regreg_mask) {
        case op_inc:
            ex_coe = true;
            ex_ar = true;
            break;
        case op_dec:
            ex_coe = false;
            ex_as = true;
            break;
        case op_move:
            ex_ar = true;
            ex_coe = false;
            ex_cr = true;
            ex_or = Or;
            break;
        case op_load:
            ex_mr = true;
            break;
        }

        // Ad. hoc. insns.
        switch (opcode) {
        case 0x40:
        case 0x44:
            ex_out = true;              // OUT
            break;
        case 0x50:
        case 0x54:
            ex_in = true;               // IN
            break;
        case 0x58:
            ex_mr = mr;                 // Might be MEM, might not.
            break;
        case 0x5c:
            ex_mr = true;               // MEM
            break;
        case 0x48:
            ex_mr = mr;                 // Might be STA, might not.
            ex_mw = mw;
            break;
        case 0x4c:
            ex_mw = true;               // STA
            ex_mr = true;
            break;
        }

        // 101...... are unused...
        if ((opcode & 0xe0) == 0xa0) {
            ex_cw = cw;
            check_alu = false;
        }

        // Test for undesirable combos.
        if (cs && cr)
            errx(1, "CS and CR on %s %#02x at %i\n", tag, opcode, i);

        // Test for undesirable combos.
        if (as && ar)
            errx(1, "AS and AR on %s %#02x at %i\n", tag, opcode, i);

        if (cw != ex_cw)
            errx(1, "CW %i exp %i on %s %#02x at %i\n",
                 cw, ex_cw, tag, opcode, i);

        if (mw != ex_mw)
            errx(1, "MW %i exp %i on %s %#02x at %i\n",
                 mw, ex_mw, tag, opcode, i);

        if (mr != ex_mr)
            errx(1, "MR %i exp %i on %s %02x at %i\n",
                 mr, ex_mr, tag, opcode, i);

        if (in != ex_in)
            errx(1, "IN %i exp %i on %s %02x at %i\n",
                 in, ex_in, tag, opcode, i);

        if (out != ex_out)
            errx(1, "OUT %i exp %i on %s %02x at %i\n",
                 out, ex_out, tag, opcode, i);

        // Only check arithmetic flags on things that use the ALU.
        if (!check_alu)
            continue;

        if (as != ex_as)
            errx(1, "AS %i exp %i on %s %02x at %i\n",
                 as, ex_as, tag, opcode, i);

        if (ar != ex_ar)
            errx(1, "AR %i exp %i on %s %02x at %i\n",
                 ar, ex_ar, tag, opcode, i);

        if (cs != ex_cs)
            errx(1, "CS %i exp %i on %s %02x at %i\n",
                 cs, ex_cs, tag, opcode, i);

        if (cr != ex_cr)
            errx(1, "CR %i exp %i on %s %02x at %i\n",
                 cr, ex_cr, tag, opcode, i);

        if (coe != ex_coe)
            errx(1, "CoE %i exp %i on %s %02x at %i\n",
                 coe, ex_coe, tag, opcode, i);

        if (And != ex_and)
            errx(1, "AND %i exp %i on %s %02x at %i\n",
                 And, ex_and, tag, opcode, i);

        if (Or != ex_or)
            errx(1, "OR %i exp %i on %s %02x at %i\n",
                 Or, ex_or, tag, opcode, i);

        if (n != ex_n)
            errx(1, "N %i exp %i on %s %02x at %i\n",
                 n, ex_n, tag, opcode, i);
    }

    fprintf(stderr, "Tested: %zi\n", S.num_samples);
    return 0;
}

File: src/opdecode.rs
use crate::spice_read::SpiceRead;

pub fn opdecode(path: &String) {
    let s = SpiceRead::from_path(path, 5e-6);
    let mut count = 0;
    let mut seen = std::collections::HashSet::new();

    for [i2, i3, i4, i5, i6, i7, ij, ii2, ii3, ii4, ii5, ii6, ii7, iij, co,
         qe, cr, cs, coe, ar, as_, and, or, n, mpre,
         in_, out, mw, mr, cw] in s.extract_positive(&[
            "i2", "i3", "i4", "i5", "i6", "i7", "ij",
            "i2#", "i3#", "i4#", "i5#", "i6#", "i7#", "ij#", "co",
            "qe", "cr", "cs#", "coe#", "ar#", "as", "and", "or", "n#", "mpre#",
            "in#", "out#", "mw", "mr#", "cw#"]) {
        assert_eq!((i2, i3, i4, i5, i6, i7), (ii2, ii3, ii4, ii5, ii6, ii7));
        assert_eq!(ij, iij);

        let opcode = i2 as u8 * 4 + i3 as u8 * 8 + i4 as u8 * 16
            + i5 as u8 * 32 + i6 as u8 * 64 + i7 as u8 * 128;
        seen.insert(opcode + co as u8);

        // Test for undesirable combos.
        assert!(!cs || !cr, "CS and CR on {opcode:#04x}");
        // AR overrides AS.
        let as_ = as_ & !ar;

        let [mut ex_cr, mut ex_cs, mut ex_ar, mut ex_as] = [false; 4];
        let [mut ex_and, mut ex_or, mut ex_n, mut ex_in] = [false; 4];
        let [mut ex_out, mut ex_mr, mut ex_mw, mut ex_cw] = [false; 4];
        let mut ex_mpre = false;

        // Arithmetic & xfer ops (except loadm).
        let mut check_alu = opcode >= 0x80;

        // Instructions that write the C flag.  ALU ops (including tests).
        if opcode & 0xc0 == 0x80 {
            ex_cw = true;
        }

        // All arithemic and xfer instructions, except loadm.
        let mut ex_qe = opcode >= 0x80 && opcode & 0xcc != 0xcc;

        let mut ex_coe = false;
        if opcode < 0xc0 {
            match opcode & 0x1c {
                0x00|0x08|0x18 => ex_coe = false,           // Add, Or, Xor.
                0x04|0x0c      => ex_coe = true,            // Sub, And.
                _              => ex_coe = co,              // Adc, Sbc.
            }
            match opcode & 0x1c {
                0x00 => (),                                 // Add
                0x04 => ex_n = true,                        // Sub
                0x08 => (ex_cr, ex_or) = (true, true),      // Or
                0x0c => (ex_cs, ex_and) = (true, true),     // And
                0x10 => (),                                 // Adc
                0x14 => ex_n = true,                        // Sbc
                0x18 => ex_cr = true,                       // Xor
                0x1c => check_alu = false,                  // DNC
                _    => unreachable!(), 
            } 
        }

        match opcode & 0xcc { // Inc/Dec/Load/Loadm
            0xc0 => (ex_coe, ex_ar) = (true, true),         // Inc
            0xc4 => (ex_coe, ex_as) = (false, true),        // Dec
            // Load.  We don't care about CR and OR for this, either ADD or OR
            // will work.
            0xc8 => (ex_coe, ex_ar, ex_cr, ex_or) = (false, true, cr, or),
            // Loadm, ALU is actually don't care.
            0xcc => (check_alu, ex_mr) = (false, true),
            _    => (),
        }

        // 0x68 : Asserts MW without MR, the refresh cycle will STA?
        // 0x78 : Asserts MPRE with no other strobes, sets K=-1?
        match opcode {
            0x60 | 0x64 => ex_out = true,                   // Out
            0x68        => (ex_mr, ex_mw) = (mr, mw),       // Not assigned.
            0x6c        => (ex_mr, ex_mw) = (true, true),   // Sta
            0x70 | 0x74 => (ex_in, ex_mpre) = (true, true), // In
            0x78        => (ex_mr, ex_mpre) = (mr, mpre),   // Not assigned.
            0x7c        => (ex_mr, ex_mpre) = (true, true), // Mem
            _           => (),
        }

        // IJ overrides everything...
        if ij {
            if opcode >= 0x40 {
                continue;              // Impossible combo.
            }
            check_alu = true;
            ex_qe = true;
            ex_cw = false;
            ex_mw = false;
            ex_mr = false;
            ex_in = false;
            ex_out = false;
            ex_mpre = false;
            ex_as = false;
            ex_ar = true;
            ex_cs = false;
            ex_cr = false;
            ex_coe = true;
            ex_and = false;
            ex_or = false;
            ex_n = false;
        }
        seen.insert(opcode + co as u8 + ij as u8 * 2);

        check(qe  , ex_qe  , "QE"  , opcode, ij);
        check(cw  , ex_cw  , "CW"  , opcode, ij);
        check(mw  , ex_mw  , "MW"  , opcode, ij);
        check(mr  , ex_mr  , "MR"  , opcode, ij);
        check(in_ , ex_in  , "IN"  , opcode, ij);
        check(out , ex_out , "OUT" , opcode, ij);
        check(mpre, ex_mpre, "MPRE", opcode, ij);

        if !check_alu {
            continue;
        }

        check(as_, ex_as , "CW" , opcode, ij);
        check(ar , ex_ar , "CR" , opcode, ij);
        check(cs , ex_cs , "CS" , opcode, ij);
        check(cr , ex_cr , "CR" , opcode, ij);
        check(coe, ex_coe, "CoE", opcode, ij);
        check(and, ex_and, "AND", opcode, ij);
        check(or , ex_or , "OR" , opcode, ij);
        check(n  , ex_n  , "N"  , opcode, ij);

        count += 1;
    }
    println!("Tested {count} of {}", s.num_samples());
    assert_eq!(seen.len(), 128 + 32);
}

fn check(got: bool, ex: bool, tag: &str, opcode: u8, ij: bool) {
    assert_eq!(got, ex,
               "got {got} expect {ex} for {tag} on {opcode:#04x} {ij}");
}
